File: data_structure/queue/src/QueueDemo.py
# Python
# データ構造: キュー (Queue)

class QueueData:
    def __init__(self):
        self._data = []
    
    def get(self):
        # 要素を取得
        return self._data

    def get_index(self, item):
        # キュー内に指定した要素があるか検索
        try:
            index = self._data.index(item)
            return index
        except ValueError:
            print(f"ERROR: {item} は範囲外です")
            return -1

    def get_value(self, index):
        # 指定したインデックスの要素を取得
        if 0 <= index < len(self._data):
            return self._data[index]
        else:
            print(f"Error: インデックス {index} は範囲外です")
            return None

    def enqueue(self, item):
        # キューの末尾に要素を追加
        self._data.append(item)
        return True
    
    def dequeue(self):
        # キューが空でない場合、先頭要素を取り出す
        if not self.is_empty():
            self._data.pop(0)
            return True
        else:
            print("ERROR: キューが空です")
            return False
    
    def peek(self):
        # キューが空でない場合、先頭要素を参照
        if not self.is_empty():
            return self._data[0]
        else:
            print("ERROR: キューが空です")
            return None
    
    def is_empty(self):
        # キューが空かどうかを確認
        return len(self._data) == 0
    
    def size(self):
        # キューの要素数を返す
        return len(self._data)
    
    def clear(self):
        # キューをクリア
        self._data = []
        return True

def main():
    print("Queue TEST -----> start")

    print("\nnew")
    queue_data = QueueData()
    print(f"  現在のデータ: {queue_data.get()}")

    print("\nis_empty")
    output = queue_data.is_empty()
    print(f"  出力値: {output}")

    print("\nenqueue")
    input = [10, 20, 30]
    for item in input:
        print(f"  入力値: {item}")
        output = queue_data.enqueue(item)
        print(f"  出力値: {output}")
        print(f"  現在のデータ: {queue_data.get()}")

    print("\nsize")
    output = queue_data.size()
    print(f"  出力値: {output}")

    print("\npeek")
    output = queue_data.peek()
    print(f"  出力値: {output}")

    print("\nget_index")
    input = 20
    print(f"  入力値: {input}")
    output = queue_data.get_index(input)
    print(f"  出力値: {output}")

    print("\nget_index")
    input = 50
    print(f"  入力値: {input}")
    output = queue_data.get_index(input)
    print(f"  出力値: {output}")

    print("\ndequeue")
    output = queue_data.dequeue()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {queue_data.get()}")

    print("\ndequeue")
    output = queue_data.dequeue()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {queue_data.get()}")

    print("\nsize")
    output = queue_data.size()
    print(f"  出力値: {output}")

    print("\ndequeue")
    output = queue_data.dequeue()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {queue_data.get()}")

    print("\ndequeue")
    output = queue_data.dequeue()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {queue_data.get()}")

    print("\nis_empty")
    output = queue_data.is_empty()
    print(f"  出力値: {output}")

    print("\nclear")
    output = queue_data.clear()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {queue_data.get()}")

    print("\nsize")
    output = queue_data.size()
    print(f"  出力値: {output}")

    print("\nQueue TEST <----- end")

if __name__ == "__main__":
    main()

File: data_structure/queue/src/QueueDemo.rs
// Rust
// データ構造: キュー (Queue)

use std::collections::VecDeque;

struct QueueData<T> {
    data: VecDeque<T>,
}

impl<T: Clone + std::fmt::Debug + PartialEq> QueueData<T> {
    fn new() -> Self {
        QueueData {
            data: VecDeque::new(),
        }
    }

    fn get(&self) -> Vec<T> {
        self.data.iter().cloned().collect()
    }

    fn get_index(&self, item: &T) -> i32 {
        match self.data.iter().position(|x| x == item) {
            Some(index) => index as i32,
            None => {
                println!("ERROR: {:?} is out of range", item);
                -1
            }
        }
    }

    fn get_value(&self, index: usize) -> Option<T> {
        if index < self.data.len() {
            self.data.get(index).cloned()
        } else {
            println!("Error: Index {} is out of range", index);
            None
        }
    }

    fn enqueue(&mut self, item: T) -> bool {
        self.data.push_back(item);
        true
    }

    fn dequeue(&mut self) -> bool {
        if !self.is_empty() {
            self.data.pop_front();
            true
        } else {
            println!("ERROR: Queue is empty");
            false
        }
    }

    fn peek(&self) -> Option<T> {
        if !self.is_empty() {
            self.data.front().cloned()
        } else {
            println!("ERROR: Queue is empty");
            None
        }
    }

    fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    fn size(&self) -> usize {
        self.data.len()
    }

    fn clear(&mut self) -> bool {
        self.data.clear();
        true
    }
}

fn main() {
    println!("Queue TEST -----> start");

    println!("\nnew");
    let mut queue_data = QueueData::<i32>::new();
    println!("  現在のデータ: {:?}", queue_data.get());

    println!("\nis_empty");
    let output = queue_data.is_empty();
    println!("  出力値: {}", output);

    println!("\nenqueue");
    let input = vec![10, 20, 30];
    for item in input {
        println!("  入力値: {}", item);
        let output = queue_data.enqueue(item);
        println!("  出力値: {}", output);
        println!("  現在のデータ: {:?}", queue_data.get());
    }

    println!("\nsize");
    let output = queue_data.size();
    println!("  出力値: {}", output);

    println!("\npeek");
    let output = queue_data.peek();
    println!("  出力値: {:?}", output);

    println!("\nget_index");
    let input = 20;
    println!("  入力値: {}", input);
    let output = queue_data.get_index(&input);
    println!("  出力値: {}", output);

    println!("\nget_index");
    let input = 50;
    println!("  入力値: {}", input);
    let output = queue_data.get_index(&input);
    println!("  出力値: {}", output);

    println!("\ndequeue");
    let output = queue_data.dequeue();
    println!("  出力値: {}", output);
    println!("  現在のデータ: {:?}", queue_data.get());

    println!("\ndequeue");
    let output = queue_data.dequeue();
    println!("  出力値: {}", output);
    println!("  現在のデータ: {:?}", queue_data.get());

    println!("\nsize");
    let output = queue_data.size();
    println!("  出力値: {}", output);

    println!("\ndequeue");
    let output = queue_data.dequeue();
    println!("  出力値: {}", output);
    println!("  現在のデータ: {:?}", queue_data.get());

    println!("\ndequeue");
    let output = queue_data.dequeue();
    println!("  出力値: {}", output);
    println!("  現在のデータ: {:?}", queue_data.get());

    println!("\nis_empty");
    let output = queue_data.is_empty();
    println!("  出力値: {}", output);

    println!("\nclear");
    let output = queue_data.clear();
    println!("  出力値: {}", output);
    println!("  現在のデータ: {:?}", queue_data.get());

    println!("\nsize");
    let output = queue_data.size();
    println!("  出力値: {}", output);

    println!("\nQueue TEST <----- end");
}
